### DISCUSSION TITLE
How to pretty print a BaseModel object?

### DISCUSSION THREAD

**Discussion by TonyYanOnFire** [pydantic/pydantic#7787] [Question] [answered]
Hi.

When pprinting a BaseModel object, I got a one-line string.
```
pprint(user)
User(id=123, signup_ts=datetime.datetime(2019, 6, 1, 12, 22), friends=[1, 2, 3], name='John Doe')
```
It will be better if the outcome indented like `pprint(user.dict())`, espicially when the model is nested. Something like this:
```
User(
    id=123, 
    signup_ts=datetime.datetime(2019, 6, 1, 12, 22), 
    friends=[1, 2, 3], 
    name='John Doe'
)
```

Is there a way to accomplish this ?

**Comment by mohammad-al-zoubi** [comment]
This made it very readable to me. It's not perfect, but does the job:

```import json
object = BaseModel()
print(json.dumps(object.__dict__, indent=4))

**Reply by TonyYanOnFire** [reply]
It does,  on a small scale. But my goal here is to  "prettyly print the exact __repr__() outcome", cause including the class names in the outcome string matters for me,

**Reply by mohammad-al-zoubi** [reply]
`def pprint(user):

    dict = str(json.dumps(user.__dict__, indent=4))
    dict = user.__class__.__name__ + dict.replace('"', '').replace(':', '=').replace('{', '(').replace('}', ')')
    print(dict)
`
Output:

`User(

    id= 123,
    signup_ts= 2020-01-01T00=00=00,
    friends= [
        1,
        2,
        3
    ],
    name= John Doe
)
`

**Comment by TonyYanOnFire** [comment, marked as answer]
While digging into pydantic's source code I found the `devtools` package, which solves this problem perfectly, as the doc described [here](https://docs.pydantic.dev/1.10/usage/devtools/)

```
from datetime import datetime
from typing import List
from pydantic import BaseModel
from devtools import pprint

class Address(BaseModel):
    street: str
    country: str
    lat: float
    lng: float

class User(BaseModel):
    id: int
    name: str
    signup_ts: datetime
    friends: List[int]
    address: Address

user = User(
    id='123',
    name='John Doe',
    signup_ts='2019-06-01 12:22',
    friends=[1234, 4567, 7890],
    address=dict(street='Testing', country='uk', lat=51.5, lng=0),
)
pprint(user)
```

Output:
```
User(
    id=123,
    name='John Doe',
    signup_ts=datetime.datetime(2019, 6, 1, 12, 22),
    friends=[
        1234,
        4567,
        7890,
    ],
    address=Address(
        street='Testing',
        country='uk',
        lat=51.5,
        lng=0.0,
    ),
)
```

**Comment by spyoungtech** [comment]
As one possible alternative.... since the `__repr__` of a model produces valid Python code, you can use `black` to print it nicely. The output is not consistent, unless you tweak the line length to be very low, but it gets the job done.

```py
import black
def pprint(model: BaseModel):
    s = repr(model)
    pretty = black.format_str(s, mode=black.FileMode(line_length=40))
    print(pretty)

# ...

pprint(user)
```

```text
User(
    id=123,
    name="John Doe",
    signup_ts=datetime.datetime(
        2019, 6, 1, 12, 22
    ),
    friends=[1234, 4567, 7890],
    address=Address(
        street="Testing",
        country="uk",
        lat=51.5,
        lng=0.0,
    ),
)
```

Or when `line_length` is set to `1` it will give a consistent output, but will obviously take up a lot more vertical space:

```text
User(
    id=123,
    name="John Doe",
    signup_ts=datetime.datetime(
        2019,
        6,
        1,
        12,
        22,
    ),
    friends=[
        1234,
        4567,
        7890,
    ],
    address=Address(
        street="Testing",
        country="uk",
        lat=51.5,
        lng=0.0,
    ),
)
```
